### PR TITLE
Add F0

### PR DESCRIPTION
**Pull request by chexq09**
Add F0 for #5 .

Use [SWIPE'](https://www.researchgate.net/publication/23558616_A_sawtooth_waveform_inspired_pitch_estimator_for_speech_and_music) to calculate F0.

```
@article{camacho2008sawtooth,
  title={A sawtooth waveform inspired pitch estimator for speech and music},
  author={Camacho, Arturo and Harris, John G},
  journal={The Journal of the Acoustical Society of America},
  volume={124},
  number={3},
  pages={1638--1652},
  year={2008},
  publisher={AIP Publishing}
}
```

### CODE REVIEW

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/repository/ChartRepository.kt [hunk 1]
```diff
@@ -200,13 +217,32 @@ object ChartRepository {
         ?.let { cacheDirectory.resolve(it) }
         ?.takeIf { it.isFile }
         ?: run {
+            // TODO: modulePrefix?
```
Comment (sdercolin) [inline src/jvmMain/kotlin/com/sdercolin/vlabeler/repository/ChartRepository.kt:220]: Yes we probably need it 😂

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/repository/ChartRepository.kt [hunk 2]
```diff
@@ -231,6 +267,7 @@ object ChartRepository {
     private const val KEY_WAVEFORM = "waveform"
     private const val KEY_SPECTROGRAM = "spectrogram"
     private const val KEY_POWER_GRAPH = "power_graph"
+    private const val KEY_FUNDAMENTAL_GRAPH = "fundamental_graph"
```
Comment (sdercolin) [inline src/jvmMain/kotlin/com/sdercolin/vlabeler/repository/ChartRepository.kt:270]: I feel that we can use simpler ones like `fundamental` and `power`
(If we change the module key, the existing cache is already invalidated, so we can change this key at the same time without extra cost)

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/dialog/preferences/PreferencesPages.kt [hunk 3]
```diff
@@ -394,6 +395,164 @@ object PreferencesPages {
...
+                        update = { copy(drawReferenceLine = it) },
+                    )
+                    color(
+                        title = Strings.PreferencesChartsFundamentalColor,
+                        defaultValue = AppConf.Fundamental.DEFAULT_COLOR,
+                        select = { it.color },
+                        update = { copy(color = it) },
+                        useAlpha = true,
+                    )
+                    color(
```
Comment (sdercolin) [inline src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/dialog/preferences/PreferencesPages.kt:538]: We can move the reference line color next to the `draw reference line` and add `enabled` to it?

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/editor/ChartStore.kt [hunk 4]
```diff
@@ -139,6 +146,15 @@ class ChartStore {
                         )
                     }
                 }
+                if (sampleInfo.hasFundamental && sampleInfo.hasSpectrogram && appConf.painter.fundamental.enabled) {
```
Comment (sdercolin) [inline src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/editor/ChartStore.kt:149]: Same here. `sampleInfo.hasSpectrogram` is redundant?

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/editor/labeler/Canvas.kt [hunk 5]
```diff
@@ -253,6 +253,21 @@ private fun Chunk(
                     }
                 }
             }
+            if (sampleInfo.hasSpectrogram && sampleInfo.hasSpectrogram &&
```
Comment (sdercolin) [inline src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/editor/labeler/Canvas.kt:256]: should be hasFundamental?

**Comment by sdercolin** [comment]
You can site the references in the source code

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/io/Fundamental.kt [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+import androidx.compose.runtime.Immutable
+import com.sdercolin.vlabeler.model.AppConf
+import kotlin.math.log2
+import kotlin.math.pow
+
+/**
+ * Data class to represent a fundamental.
+ *
+ * @param data A list of fundamental data. Each element is a fundamental data.
```
[the remark this answers is not in the document]
Comment (chexq09) [reply src/jvmMain/kotlin/com/sdercolin/vlabeler/io/Fundamental.kt:11]: Done.

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/io/SWIPE.kt [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+     * @param conf The configuration of the fundamental frequency.
+     * @param sampleRate The sample rate of the wave.
+     */
+    private fun updateKernel(conf: AppConf.Fundamental, sampleRate: Float): SWIPEKernelData? {
+        if (conf.semitoneSampleNum < 1 || conf.maxFundamental < conf.minFundamental) {
+            return null
+        }
+
+        // Calculate the candidate fundamental frequency.
+        // TODO: if conf.semitoneSampleNum > 8, calculate kernel by conf.semitoneSampleNum == 8, then interpolate.
```
[the remark this answers is not in the document]
Comment (chexq09) [reply src/jvmMain/kotlin/com/sdercolin/vlabeler/io/SWIPE.kt:83]: The result is discrete frequency now. We can interpolate it to get a continuous frequency.
The TODO is moved to L277 where the interpolation should be done.

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/io/SWIPE.kt [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    // FFT
+    val absoluteSpectrogram: Array<DoubleArray> = Array(frames.size) { i ->
+        val signal = frames[i]
+        val fft = FastFourier(signal)
+        fft.transform()
+        val magnitude = fft.getMagnitude(true)
+        magnitude.copyOf((magnitude.size * maxFrequencyRate).toInt())
+    }
+
+    return if (absoluteSpectrogram.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (chexq09) [reply src/jvmMain/kotlin/com/sdercolin/vlabeler/io/SWIPE.kt:352]: Done.

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/repository/ChartRepository.kt (hunk 1 above)
Comment (chexq09) [reply src/jvmMain/kotlin/com/sdercolin/vlabeler/repository/ChartRepository.kt]: Done.

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/repository/ChartRepository.kt (hunk 2 above)
Comment (chexq09) [reply src/jvmMain/kotlin/com/sdercolin/vlabeler/repository/ChartRepository.kt]: Done.

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/dialog/preferences/PreferencesPages.kt (hunk 3 above)
Comment (chexq09) [reply src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/dialog/preferences/PreferencesPages.kt:538]: I think the `color` should be shown at first, so I move the `draw reference line`.

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/model/SampleInfo.kt [hunk 9]
```diff
@@ -62,13 +63,18 @@ data class SampleInfo(
...
     val algorithmVersion: Int,
 ) {
 
     val totalChartCount: Int
         get() = chunkCount *
-            (channels + (if (hasSpectrogram) 1 else 0) + (if (hasPower) powerChannels else 0))
+            (
+                channels + (if (hasSpectrogram) 1 else 0) +
+                    (if (hasPower) powerChannels else 0) +
+                    (if (hasFundamental && hasSpectrogram) 1 else 0)
```
[the remark this answers is not in the document]
Comment (chexq09) [reply src/jvmMain/kotlin/com/sdercolin/vlabeler/model/SampleInfo.kt]: In the first version, the fundamental is calculated by spectrogram. But now, it is calculated by wave.
So `hasSpectrogram` is not needed.

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/editor/ChartStore.kt (hunk 4 above)
Comment (chexq09) [reply src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/editor/ChartStore.kt]: Remove `sampleInfo.hasSpectrogram`.

File: src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/editor/labeler/Canvas.kt (hunk 5 above)
Comment (chexq09) [reply src/jvmMain/kotlin/com/sdercolin/vlabeler/ui/editor/labeler/Canvas.kt]: Change to `hasFundamental`.

**Pull request merged**